hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java

```diff
@@ -222,7 +223,8 @@ public Iterator<HoodieRecord> iterator() {
   }
 
   public Map<String, HoodieRecord> getRecords() {
-    return records;
+    return records.entrySet().stream().collect(
```

Comment:
Can we override this method instead?

Reply:
Since we have made the super class `records` cache generic, so we have to make changes here to return String.

Reply:
Does generic type solves the problem? Or can we do a implicit type cast.

Reply:
No type cast is required, generic type is sufficient. So, this method should always return string otherwise we have to change all the callers, including merge handle, which we don't want to do.

Reply:
> generic type is sufficient

For generic type, I mean to make the it as a Java generic type like `K extends Serializable` instead of just Serializable, does that work?

Reply:
That will also work but let me put down my thoughts why I considered to use `Serializable` as the key type directly instead of a generic type `K extends Serializable`. Firstly, if we use `K extends Serializable`, the creators of `HoodieMergedLogRecordScanner` need to specify the correct type that matches the key. This would increase the surface area of the changes.

Plus, using `Serializable` as the key type directly is more flexible as it allows the `ExternalSpillableMap` to accept any object that implements Serializable. This means you can use a wide variety of types as keys without needing to specify a particular type parameter when instantiating the map.

One drawback of using `Serializable` directly as the key type is the reduced type safety. However, this is only an internal API implementation so it is less prone to being misused.

Reply:
Discussed offline, and we are aligned on the following approach (which is also done as part of the last commit):
Introduce a new abstract class `BaseHoodieMergedLogRecordScanner` which subclasses from `AbstractHoodieLogRecordReader`. The new abstract class holds the `records` map as `ExternalSpillableMap<K, HoodieRecord>` and exposes `public abstract Map<K, HoodieRecord> getRecords()` API. The existing `HoodieMergedLogRecordScanner` now derives from the new abstract class (instead of `AbstractHoodieLogRecordReader`) and uses String keys.